### PR TITLE
Allow project search on admin dashboard

### LINKED ISSUE #7636: [Admin] Search for projects and show project details
Admin page should allow searching of projects by entering the repo or the slug.

(The following screenshot is just a placeholder - this will need more design decisions)
Screenshot 2022-01-17 at 11 13 21

#### Search

- [x] sub-string match anywhere in the cloneUrl (assuming this is possible, and OK for DB performance)
- [x] nice-to-have: match on slug 

#### Search results
- [x] list entries show name (link to project detail), cloneUrl (link to git provider)
- [x] list ordered by created date
- [ ] ~list ordered by updated date~

#### Project details
- [x] page URL is /admin/projects/project-id
- [x] name
- [x] repository (from cloneUrl)
- [x] incremental prebuilds (from settings)
- [x] owner
- [x] team
- [x] created date
- [x] list of prebuilds

#### ~Prebuild details~
- [ ] ~page URL is /admin/prebuilds/prebuild-id~
- [ ] ~same content (with prebuild log) as in projects UI~

### PR DESCRIPTION
**Pull request by laushinka**
## Description
1. Allows searching for projects
2. Shows project detail
3. Shows list of prebuilds for each project

| Search result | Detail with prebuilds | Detail without prebuilds
| - | - | - |
| Screenshot 2022-01-31 at 02 47 33 | Screenshot 2022-01-31 at 18 58 40 | Screenshot 2022-01-31 at 18 55 24 | 

## Related Issue(s)
Fixes #7636 and #5022

## How to test
1. Go to https://laushinka-search-project-7636.staging.gitpod-dev.com/admin/projects
2. To search all projects, click on the "Search" button without any search queries.
3. To search specific projects, enter the search query.
4. Once there are result(s), click on one of the rows to go to the project detail.

## Release Notes

```release-note
Admins can do a text search for projects and their associated prebuilds.
```

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **laushinka** after the PR has been reviewed.

Associated issue: *#7636*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by laushinka** [comment]
@AlexTugarev I requested your review because we talked re the Prebuilds component already, so I thought it would make sense. But feel free to let me know in case your plate is full and I'll assign someone else 🤗

**Comment by gtsiolis** [comment]
Looking at this now! 👀

**Comment by jldec** [comment]
Trying out the project search I noticed that the state of the search input is lost when I click through to a project detail and then go back to the project list using the back button.

This is a bit confusing because searching with no input the first time, returns a full list, but clicking the search button again after coming back from a project detail, appears to do nothing (the previous list remains unchanged).

**Comment by jldec** [comment]
Q: Were you able validate that project searches won't trigger a full DB table scan (or if they do, it's ok when that happens)?

**Review by jldec — CHANGES_REQUESTED** [review]
This looks really good - thanks @laushinka 
I left 1 comment about search state on back-button, and a question about DB table scans.

File: components/dashboard/src/admin/ProjectDetail.tsx
Comment (gtsiolis) [inline components/dashboard/src/admin/ProjectDetail.tsx:29]: **suggestion**: What do you think about merging these two columns into one and include a subtle identifier for differentiating team or user ownership? This was we could drop the extra empty column and maybe also reorder the elements so that secondary information like incremental prebuilds or soft deletion downwards. For example:

| BEFORE | AFTER |
|-|-|
| project-before | project-after |

File: components/dashboard/src/projects/Prebuilds.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/Prebuilds.tsx:236]: **praise**: Thanks for including an empty state here, @laushinka! 🌟

**Comment by laushinka** [comment]
> Q: Were you able validate that project searches won't trigger a full DB table scan (or if they do, it's ok when that happens)?

I created an index for the cloneUrl so it shouldn't do a full table scan.

File: components/dashboard/src/projects/Prebuilds.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/Prebuilds.tsx:236]: **suggestion**: Since you added an empty state here, could we adjust the spacing here to provide some visual balance? You could borrow the proposal layout from https://github.com/gitpod-io/gitpod/issues/5022. 🏓

File: components/dashboard/src/admin/ProjectsSearch.tsx
Comment (gtsiolis) [inline components/dashboard/src/admin/ProjectsSearch.tsx:122]: **praise**: Thanks for adding all the truncation classes here! 💯

File: components/dashboard/src/admin/ProjectsSearch.tsx
Comment (gtsiolis) [inline components/dashboard/src/admin/ProjectsSearch.tsx:120]: **suggestion**: Since this was also discussed for the upcoming new workspace input modal (https://github.com/gitpod-io/gitpod/pull/7715), we could re-use the same pattern which is non-existing yet to include a small indicator on hover that describes the action of the single-click open (see screenshot below). However, feel free to ignore this comment for now, bringing this up for visibility and future iterations.

| Open workspace | View project |
|-|-|
| Modal copy | Projects (List) |

File: components/dashboard/src/projects/Prebuilds.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/Prebuilds.tsx:191]: **praise**: Nice catch! I was just about to check this out. ⚾

File: components/dashboard/src/admin/ProjectsSearch.tsx
Comment (gtsiolis) [inline components/dashboard/src/admin/ProjectsSearch.tsx:100]: **issue(non-blocking)**: The issue about state handling that @jldec mentioned in https://github.com/gitpod-io/gitpod/pull/7882#issuecomment-1025706941 is valid but also visible when searching users or workspaces. If this is an online-fix let's include it here but probably not worth blocking this PR as it seems to be out of the scope of the changes here. 🏀

**Review by gtsiolis — APPROVED** [review]
Thanks for adding this, @laushinka! 🔮 

Left below some non-blocking comments about UX changes.

I also agree with @jldec's comment in https://github.com/gitpod-io/gitpod/issues/7636#issuecomment-1024096006 about merging the first iteration of this without further changes. ✔️ 

Approving to unblock merging but holding in case someone from the team could take a closer look at the code changes. 🏀 

/hold

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: bb25ac53ac55891dbca1bd3bda6355d5d29809c6</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7882#pullrequestreview-867809190" title="Approved">gtsiolis</a>*

Associated issue: *#7636*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [gtsiolis]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [gtsiolis]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
New changes are detected. LGTM label has been removed.

File: components/dashboard/src/projects/Prebuilds.tsx
Comment (laushinka) [reply components/dashboard/src/projects/Prebuilds.tsx:191]: Ended up removing this check for #5022.

File: components/dashboard/src/projects/Prebuilds.tsx
Comment (laushinka) [reply components/dashboard/src/projects/Prebuilds.tsx]: Done!

File: components/dashboard/src/admin/ProjectDetail.tsx
Comment (laushinka) [reply components/dashboard/src/admin/ProjectDetail.tsx]: Done

File: components/dashboard/src/admin/ProjectsSearch.tsx
Comment (laushinka) [reply components/dashboard/src/admin/ProjectsSearch.tsx:120]: Makes sense!

File: components/dashboard/src/admin/ProjectDetail.tsx
Comment (gtsiolis) [inline components/dashboard/src/admin/ProjectDetail.tsx:35]: **issue**: This is minor but the column alignment seems slightly off for this element. We probably don't need the `ml-3` class here or the first child elements for each row, at least until we start working on more complex responsive layouts. What do you think?

| BEFORE | AFTER |
|-|-|
| before | after |

File: components/dashboard/src/admin/ProjectDetail.tsx
Comment (gtsiolis) [inline components/dashboard/src/admin/ProjectDetail.tsx:28]: **suggestion**: What do you think of removing the type indicator (user/Team) from the hyperlink and using a softer color here? The `text-gray-400` should suffice for the light theme. For the dark theme we probably need to update the text colors for better hierarchy as we did in https://github.com/gitpod-io/gitpod/issues/6762 for integrations.

**Review by gtsiolis — CHANGES_REQUESTED** [review]
Thanks for the udpate, @laushinka! 🔮 

UX LGTM. Left some minor comments for the visuals.

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **gtsiolis** after the PR has been reviewed.

Associated issue: *#7636*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/dashboard/src/admin/ProjectsSearch.tsx
Comment (laushinka) [reply components/dashboard/src/admin/ProjectsSearch.tsx:119]: Ah, I removed it because I thought that was the preferred way. Sure, will put it back!

File: components/dashboard/src/admin/ProjectDetail.tsx
Comment (laushinka) [reply components/dashboard/src/admin/ProjectDetail.tsx]: Done

File: components/dashboard/src/admin/ProjectsSearch.tsx
Comment (laushinka) [reply components/dashboard/src/admin/ProjectsSearch.tsx:120]: Done!

File: components/dashboard/src/admin/ProjectDetail.tsx
Comment (laushinka) [reply components/dashboard/src/admin/ProjectDetail.tsx]: I just did it in a hacky way and `markedDeleted` is still off 😅 Feel free to share your CSS magic with me �
Screenshot 2022-01-31 at 22 48 55
�

File: components/dashboard/src/admin/ProjectsSearch.tsx
Comment (jldec) [reply components/dashboard/src/admin/ProjectsSearch.tsx:100]: ok - agree.

**Review by jldec — APPROVED** [review]
Very nice work - I like this.

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: ea92d1cd6ef722adc601143a1a34ac73dcf73332</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7882#pullrequestreview-868701205" title="Approved">jldec</a>*

Associated issue: *#7636*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [jldec]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [jldec]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [jldec]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [jldec]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by jldec** [comment]
@gtsiolis I think this is ready to merge.

**Comment by laushinka** [comment]
> I think this is ready to merge.

@jldec I think you can remove the hold if you want. I haven't had any devs look at the code though. Maybe I'll ping @JanKoehnlein for a look if the others are busy?

File: components/dashboard/src/admin/ProjectsSearch.tsx
Comment (gtsiolis) [inline components/dashboard/src/admin/ProjectsSearch.tsx:123]: **suggestion(non-blocking)**: What do you think of toning down the font weight and font color here to help with information hierarchy as we do with other similar lists?

```suggestion
                        <div className="text-gray-500 dark:text-gray-100 truncate">{p.project.cloneUrl}</div>
```

| BEFORE | AFTER |
|-|-|
| list-before | list-after |

**Review by gtsiolis — COMMENTED** [review]
@laushinka I've left one last minor comment regarding UX.

Other than that, UX looks good to merge. 🏁 

I'd suggest to get at least one pair of eyes on the code changes before releasing the hold.

**Comment by roboquat** [comment]
New changes are detected. LGTM label has been removed.

**Comment by JanKoehnlein** [comment]
LGTM

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: f44654c76a2dc080c1977a892f947ab687ac7561</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7882#pullrequestreview-869173620" title="Approved">JanKoehnlein</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/7882#pullrequestreview-868701205" title="Approved">jldec</a>*

Associated issue: *#7636*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [JanKoehnlein,jldec]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [JanKoehnlein,jldec]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [JanKoehnlein,jldec]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [JanKoehnlein,jldec]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/dashboard/src/admin/ProjectDetail.tsx
Comment (laushinka) [reply components/dashboard/src/admin/ProjectDetail.tsx]: Sorry I think your comment came at the same time as the merging! Will do this as a drive-by in the next admin ticket 🙏🏽

Comment (gtsiolis) [reply components/dashboard/src/admin/ProjectDetail.tsx]: Thanks, @laushinka! Took a bias-for-action and opened https://github.com/gitpod-io/gitpod/pull/7960 for this minor fix.

**Pull request merged**
